core/node/node_sync/src/batch_transaction_updater/mod.rs

```diff
@@ -0,0 +1,302 @@
+            // its finality status. This value is untrusted as seen_at_block may have been cached
+            let Some(finality_update) = l1_block_numbers
+                .get_finality_update(db_eth_tx_history.eth_tx_finality_status, sent_at_block)
+            else {
+                continue;
+            };
+
+            // we can potentially update finality status, but we need to
+            // validate all properties including block number as it may
+            // have been cached
+            let receipt = match receipt {
+                Some(receipt) => receipt,
+                None => {
+                    let Some(receipt) =
+                        self.sl_client.tx_receipt(db_eth_tx_history.tx_hash).await?
+                    else {
+                        tracing::warn!(
+                            "Expected transaction {} to be mined at block {}, but transaction not mined at all on SL",
+                            db_eth_tx_history.tx_hash,
+                            db_eth_tx_history.sent_at_block.unwrap()
+                        );
+                        connection
+                            .eth_sender_dal()
+                            .unset_sent_at_block(db_eth_tx_history.id)
+                            .await?;
+                        continue;
+                    };
+                    receipt
+                }
+            };
+
+            let db_eth_history_id = db_eth_tx_history.id;
+            let result = self
+                .apply_status_update(
+                    &mut connection,
+                    db_eth_history_id,
+                    receipt,
+                    &l1_block_numbers,
+                )
+                .await;
+
+            if result.is_ok() {
+                updated_count += 1;
+            } else {
+                tracing::warn!(
+                    "Failed to update finality status for transaction {} with type {} from {:?} to {:?} due to error: {}",
+                    db_eth_tx_history.tx_hash,
+                    db_eth_tx_history.tx_type,
+                    db_eth_tx_history.eth_tx_finality_status,
+                    finality_update,
+                    result.err().unwrap()
+                );
+            }
+        }
+        Ok(updated_count)
+    }
+
+    pub async fn loop_iteration(
+        &self,
+        sl_chain_id: SLChainId,
+        l1_block_numbers: L1BlockNumbers,
+    ) -> anyhow::Result<i32> {
+        let mut connection = self
+            .pool
+            .connection_tagged("batch_transaction_updater")
+            .await?;
+        let to_process: Vec<TxHistory> = connection
+            .eth_sender_dal()
+            .get_unfinalized_tranasctions(self.processing_batch_size, Some(sl_chain_id))
+            .await?;
+
+        if to_process.is_empty() {
+            // Check if there are any unfinalized transactions for other chain_ids.
+            // If there are, we need to restart to make the node change the SL chain_id.
```

Comment:
Dumb question: What if `to_process` is *not* empty, but there are still non-finalized transaction for other chains? Is this normal? Asking because a way to avoid race conditions mentioned below is to load non-finalized transactions for all chains in any case instead of two-step loading.

Reply:
The idea behind this crashing is to properly process transactions around gateway migration. We get from DB transactions matching the current "mode" (SL network we are connected to). If we process all, this signifies we might be at a migration border. We don't want to restart/crash before we process all transactions before migration, we actually determine which network to "connect" to based on the oldest batch with unfinalized transnctions.

An alternative would be to get unfinalized transactions sorted by batch number, but this would require a more expensive query.

Reply:
I see; for some reason, I thought `to_process` was ordered by L1 batch number already.